src/wasm/wasm-binary.cpp

```diff
@@ -2818,6 +2853,17 @@ void WasmBinaryReader::readNextDebugLocation() {
 
     int32_t positionDelta = readBase64VLQ(*sourceMap);
     uint32_t position = nextDebugLocation.availablePos + positionDelta;
+
+    nextDebugLocation.previousPos = nextDebugLocation.availablePos;
+    nextDebugLocation.availablePos = position;
```

Comment:
This is not new logic - just moved up from below (since we also want it for a 1-length entry).

Reply:
@kripken 
Hi Alon,
does this optimization only impact the debug info read process? how about the `BinaryenFunctionSetDebugLocation` API? I think AssemblyScript is using this API for debug info generation, it also have the smeared issue for the shadow stack check.

Reply:
This part of the code, or this PR as a whole?

The PR as a whole affects binary read and write. We now write 1-length segments to prevent smearing, and we read them properly as well.

`SetDebugLocation` should still be fine. That sets debug info on a single expression. This PR fixes things in binary reading and writing, which due to smearing would cause things to get debug info when they should not. See the testcase in the PR for an example of a bug that is fixed.